mapview: add minimap for easy navigation

danix800 opened the pull request:
(https://github.com/freeplane/freeplane/pull/195)

This is a basically usable version of minimap I can come up with since I'm unfamiliar with freeplane nor swing. Bugs are expected. Please give advices.

The idea is borrowed from here: [MiniMap](https://github.com/aterai/java-swing-tips/blob/master/MiniMap/src/java/example/MainPanel.java)

Supported features:

- mouse wheel (zoom)
- mouse click/dragged (quick jump to)
- control + M (toggle on/off)
- use fixed minimap size (configurable in Preference -> Appearence -> LAF)

The widgets relationship (dashed line) is changed to the below (green dashed region):

![Screenshot from 2021-05-16 23-55-35](https://user-images.githubusercontent.com/1921413/118440786-c362d900-b71a-11eb-9dcb-a2830b5ce48e.png)

Commit:
mapview: add minimap for easy navigation

danix800 commented:
The idea is borrowed from here: https://github.com/aterai/java-swing-tips/blob/master/MiniMap/src/java/example/MainPanel.java

dpolivaev commented:
Thanks, I am going to look at your pull request at the weekend.
Regards, Dimitry

Commit:
check against extreme cases (zero size dimension)

Commit:
switch wheel direction mapping

forward  => zoom in
backward => zoom out

Commit:
fix minimap thumbnail algorithm

1. Translate and scale origianl map view into a square image;
2. Translate and scale the inner bounds, expand it to a square;
3. Then crop the image with the square;
4. Scale the square image to fixed size.

Commit:
make minimap size configurable

Commit:
confine minimap zoom level

Commit:
fix minimap thumbnail algorithm

Commit:
Fix minimap viewport calculation & mouse event processing

Commit:
fix image scaling algorithm & check against image out-of-bounds

Commit:
set minimap magnifier cursor

Commit:
properly set cursor hotspot

Commit:
add map changing listener for minimap update

Commit:
Paint directly into minimap with proper transform

Eliminate intermediate image (save memory usage)

Commit:
optimize minimap painting further

1. only update minimap if visible
2. create minimap ImageIcon only once, redraw on its graphics if
   updating is needed

Commit:
fix minimap minor issues

1. slow down wheel zoom a bit
2. display zoom level in fullscreen mode (SW corner)
3. update minimap border color value from event.NewValue
   rather than from MapView.getBackground() since view
   is not updated yet

Commit:
fix minimap wheel zoom bounding algorithm

Commit:
minor fix minimap position

Commit:
fix minimap MapChangeListener add/remove

Commit:
switch mapview before scrolling over minimap

Commit:
minimap responds to accelerator (ctrl+m on/off) in fullscreen mode

dpolivaev commented on freeplane/src/external/resources/xml/preferences.xml:
```diff
@@ -94,6 +94,7 @@
 				<number name="max_menu_item_count" min="10" />
 				<number name="menuItemFontSize" min="3" />
 				<length name="toolbar_icon_height" defaultUnit = "pt" min="0" max = "1000" step = "0.1"/>
+				<length name="minimap_size" defaultUnit = "px" min="240" max = "640" step = "1"/>
```
I would rather use points and not pixels for default units because they are adjusted to monitor resolution

danix800 replied:
Agreed and I'll change `px` to `pt`.

dpolivaev commented on freeplane/src/external/resources/xml/filemodemenu.xml:
```diff
@@ -105,6 +106,7 @@
 		<Entry name="ToggleToolbarAction" />
 		<Entry name="ShowFilterToolbarAction" />
 		<Entry name="ToggleScrollbarsAction" />
+		<Entry name="ToggleMinimapAction" accelerator="control M" />
```
I would rather not assign this action a default accelerator in any mode because accelerators are limited and similar actions do not have a default accelerator.

danix800 replied:
I didn't even notice the existing great user-definable key mapping management feature. I'll remove this.

dpolivaev commented on freeplane/src/viewer/resources/freeplane.properties:
```diff
@@ -540,6 +540,7 @@ neverShowSurvey=false
 connector_arrows=FORWARD
 connector_dash=SOLID
 toolbar_icon_height=16 pt
+minimap_size=320 px
```
You could use points, not pixels and convert points to pixels using org.freeplane.api.Quantity.toBaseUnitsRounded()

danix800 replied:
Same as above I'll change `px` to `pt`.

dpolivaev commented on freeplane/src/main/java/org/freeplane/view/swing/map/MiniMapView.java:
```diff
@@ -0,0 +1,343 @@
+import org.freeplane.features.ui.IMapViewManager;
+import org.freeplane.features.ui.ViewController;
+
+public class MiniMapView extends JPanel implements IFreeplanePropertyListener, IMapChangeListener {
```
Have you considered placing the minimap inside a separate non-modal dialog so that it can be easily resized and positioned at any monitor and at any place of the desktop not limited to the current frame?

danix800 replied:
I don't know for sure.

I searched for possible mini-map implementations in swing and the existing component hierarchy
came to me first, so I tried for several iterations and finally this PR.

It's one of the typical looks of mini-map, while it's also very attractive to use resizable and arbitrary positioned dialog. I'm acceptable with this existing implementation because the mini-map can be easily switched on/off (that's the reason I added the control-M mapping at the beginning).

I don't know how the public users would choose.

dpolivaev replied:
I agree with you.
Let me know when you finish your work and I can merge it.

Commit:
remove minimap toggle action key mapping

Commit:
use pt instead of px for minimap size unit

dpolivaev commented:
I am thinking about the best name for the new feature and for the new control. Currently it's called minimap. I can also think about mini map, map overview, bird's eye view. Which one do you like the most?

ldmpub commented:
Hi @dpolivaev I personnaly like:
.1 « minimap » (in 1 word or «  mini-map ») => word used from the begining
.2 « bird's eye » (or « aerial view » ?) is fine also according to me.

ldmpub commented:
2 shorts questions related to this new feature:

1. when would it be available for test in Freeplane Preview?
2. how documentation should be handled?

Thanks for your answers.

dpolivaev commented:
All decisions are made by @danix800, i am fine with how it goes. He can decide which words to use, I just share some thoughts. And he can decide when I should merge it to the development branch 1.9.x and then I can create a preview.
I don't know who wants to care about the documentation, but the feature seems to be very easy to use intuitively.

danix800 commented:
Minimap is the first thing flashed into my mind. But I might not use it anymore.

I thought about the feature naming over a few iterations since @dpolivaev mentioned it. The option "Overview" is more reasonable to me. The minimap is not a newly created thing to mindmap world (People might take it as a new thing at first glance), it is just an overview of the map they created.

After I thought about the name "overview" I crosschecked with XMind, and yes they call it overview.

This is the last thing remained to discuss. If we can agree on the naming then it's ok to be merged.

How do you think? @dpolivaev @ldmpub

Commit:
renaming MiniMapView to MapOverview

danix800 commented:
I made one more commit for the renaming so that you can test it easily.

If that's ok I'll stick to this name.

ldmpub commented:
"overview" is perfect for me 🙏=> we can easily tell translators to have a look to XMind to get answer to their question: "How can I translate this new Freeplane feature in my language?" 😉. Waiting Weblate to be updated to do my part 🚧.

dpolivaev commented:
I merged the pull request to make progress on this issue.
I replaced the text "Map Overview" with "Map overview" for consistency.
I removed the option for the map overview size from the preferences because of two reasons:

1. Freeplane had to be restarted for this option to take effect, I thought it is not user-friendly enough
2. I would like to change this setting using drag-and-drop at the upper left corner (about 5-10% of the overview size) of the overview which could immediately have an effect and also write the updated value to the preferences. To indicate it I would change the mouse cursor accordingly when the mouse is inside this region. I would appreciate if you could implement it, otherwise, I have to do it myself in the future.

Probably it is also time to think about refactoring: currently MapOverview is a component containing mapOverviewPanel and MapViewScrollPane. I would probably at least rename it to MapViewPane because the actual overview is the panel it contains.
How do you see it?

dpolivaev commented:
I have also fixed how this component handles when Freeplane goes to fullscreen and back.

dpolivaev commented:
The feature is available in preview 1.9.3_01
I have seen that it works less nice under MacOS because there graphics can have scale different from 1 and therefore the captured image looks pixeled. I am going to fix it later.

Anyway I appreciate your contribution very much and I would be glad if you could contribute more amazing stuff.

danix800 commented:
Really glad to see this, and thanks for your proposals, suggestions and nice fixes!

danix800 commented:
> I merged the pull request to make progress on this issue.
> I replaced the text "Map Overview" with "Map overview" for consistency.
> I removed the option for the map overview size from the preferences because of two reasons:
> 
> 1. Freeplane had to be restarted for this option to take effect, I thought it is not user-friendly enough
> 2. I would like to change this setting using drag-and-drop at the upper left corner (about 5-10% of the overview size) of the overview which could immediately have an effect and also write the updated value to the preferences. To indicate it I would change the mouse cursor accordingly when the mouse is inside this region. I would appreciate if you could implement it, otherwise, I have to do it myself in the future.
> 
> Probably it is also time to think about refactoring: currently MapOverview is a component containing mapOverviewPanel and MapViewScrollPane. I would probably at least rename it to MapViewPane because the actual overview is the panel it contains.
> How do you see it?



> I merged the pull request to make progress on this issue.
> I replaced the text "Map Overview" with "Map overview" for consistency.
> I removed the option for the map overview size from the preferences because of two reasons:
> 
> 1. Freeplane had to be restarted for this option to take effect, I thought it is not user-friendly enough
> 2. I would like to change this setting using drag-and-drop at the upper left corner (about 5-10% of the overview size) of the overview which could immediately have an effect and also write the updated value to the preferences. To indicate it I would change the mouse cursor accordingly when the mouse is inside this region. I would appreciate if you could implement it, otherwise, I have to do it myself in the future.
> 
> Probably it is also time to think about refactoring: currently MapOverview is a component containing mapOverviewPanel and MapViewScrollPane. I would probably at least rename it to MapViewPane because the actual overview is the panel it contains.
> How do you see it?

Will see what I can do :ok

danix800 commented:
I'll consider implementing the overview as a dialog when resizing is needed in future, for now I'll spend a bit of time to learn more basics of freeplane. https://github.com/freeplane/freeplane/pull/202 is another feature that I want to discuss in the future, if you have time please share your thoughts on this, for now I'll use it locally for easy navigation when I have only keyboard on my laptop.

Thanks.

dpolivaev commented:
I had to fix some other stuff, see https://github.com/freeplane/freeplane/commit/9faad789fc35ddf59f9f202f87ebceec31e91c4a

dpolivaev commented:
I have also renamed the class to MapViewPane in some later commit

dpolivaev commented:
Could you please replace the zoom-in cursor by 32x32 pixel image? 
Current 16x16 version looks not smooth enough, consider https://sourceforge.net/p/freeplane/discussion/758437/thread/494dfaa629/#a5c4

dpolivaev commented:
How can I help you in relation to https://github.com/freeplane/freeplane/pull/202 ?

danix800 commented:
> Could you please replace the zoom-in cursor by 32x32 pixel image?
> Current 16x16 version looks not smooth enough, consider https://sourceforge.net/p/freeplane/discussion/758437/thread/494dfaa629/#a5c4

I scaled the original image for several times but all the results are not perfect, the original is posted here:

![magnifier](https://user-images.githubusercontent.com/1921413/119374405-1f65c880-bcec-11eb-90aa-7da32791c8ce.png)

32x32: 

![mag32](https://user-images.githubusercontent.com/1921413/119375106-fabe2080-bcec-11eb-853f-3e354a5e78b3.png)

Or maybe these ones (from [GIMP](https://gitlab.gnome.org/GNOME/gimp) repo)

![cursor-zoom](https://user-images.githubusercontent.com/1921413/119374658-6bb10880-bcec-11eb-9693-9456c6397cc0.png)
![cursor-zoom-x2](https://user-images.githubusercontent.com/1921413/119374663-6c499f00-bcec-11eb-93c1-db6e39dd3185.png)
![mag16-1](https://user-images.githubusercontent.com/1921413/119374666-6ce23580-bcec-11eb-866e-30b3b384331a.png)
![mag24](https://user-images.githubusercontent.com/1921413/119374667-6ce23580-bcec-11eb-9331-44a22060fcdc.png)

I don't know if we can use these directly. Or maybe we can use predefined `CROSSHAIR_CURSOR` if no other good choice is available.

If we can decide on the option, I'll do the commit.

danix800 commented:
> I have also renamed the class to MapViewPane in some later commit

Thanks for all the fixes. Hope this PR does not cause too much trouble. I'll stop commit any new code before I get some real understanding of the internals of this repo.

danix800 commented:
> How can I help you in relation to #202 ?

I want to know whether this kind of code is acceptable :-) I saw someone asking for easy keyboard navigation and I'm one of them. I also searched for posts in the community, someone mentioned that scripting API might be one option for this, for now I know nothing about scripts in freeplane. I think you might know the better places or directions for this kind of features.

As for me, I think #202 is too much hard-coded so I hesitated and closed #202 .

danix800 commented:
This is how `CROSSHAIR_CURSOR` looks on my machine:

![Screenshot from 2021-05-25 00-54-34](https://user-images.githubusercontent.com/1921413/119384084-c0598100-bcf6-11eb-8c8b-21eab19c643b.png)

danix800 commented:
This is another map tool which uses `CROSSHAIR_CURSOR`:

![Screenshot from 2021-05-25 01-46-20](https://user-images.githubusercontent.com/1921413/119387255-42e43f80-bcfb-11eb-94f9-df16b26c8525.png)

XMind uses `HAND_CURSOR`.

dpolivaev commented:
> > I have also renamed the class to MapViewPane in some later commit
> 
> Thanks for all the fixes. Hope this PR does not cause too much trouble. I'll stop commit any new code before I get some real understanding of the internals of this repo.

Please don't stop. Your contributions have been valuable. Everyone can make mistakes living in unperfect world and dealing with unperfect code.
I believe that ping-pong, collaboration, communication is a better way to move forward and to avoid mistakes, the more and the earlier the better. Let us try to make the most out of your time. If you get any questions to the code, build, debugging or anything else or want to try pair programming with me just le me know.

dpolivaev commented:
And it is not just about mistakes. In my experience, many ideas and refactorings can be discovered as a result of communication.

dpolivaev commented:
Related to the cursor, I don't want to decide which solution is the best. I could ask someone to create a new cursor image looking similar to yours. We could use the CROSSHAIR_CURSOR instead. We can use it for the moment and replace it by something else later. I would appreciate a pull request. 

Where do you have your original cursor image from?

dpolivaev commented:
> > How can I help you in relation to #202 ?
> 
> I want to know whether this kind of code is acceptable :-) I saw someone asking for easy keyboard navigation and I'm one of them. I also searched for posts in the community, someone mentioned that scripting API might be one option for this, for now I know nothing about scripts in freeplane. I think you might know the better places or directions for this kind of features.
> 
> As for me, I think #202 is too much hard-coded so I hesitated and closed #202 .

Well, actually it was hard-coded because the (pretty old) original code is also hard-coding this logic. 
And I do not apply the same high code quality criteria to the old code as to my newest and latest projects. Because life is too short to refactor everything.

So I would actually suggest that we have a conversation about it. If you prefer we can continue to discuss everything here, but I would also be interested to have video chat with screen sharing. I am generally available on Fridays, Saturdays and Sundays, if it sounds good for you.

danix800 commented:
> Related to the cursor, I don't want to decide which solution is the best. I could ask someone to create a new cursor image looking similar to yours. We could use the CROSSHAIR_CURSOR instead. We can use it for the moment and replace it by something else later. I would appreciate a pull request.
> 
> Where do you have your original cursor image from?

The original magnifier image is downloaded from google image search, there could be copyright issues, I will stop using it.

See https://github.com/freeplane/freeplane/pull/203.

danix800 commented:
> > > How can I help you in relation to #202 ?
> > 
> > 
> > I want to know whether this kind of code is acceptable :-) I saw someone asking for easy keyboard navigation and I'm one of them. I also searched for posts in the community, someone mentioned that scripting API might be one option for this, for now I know nothing about scripts in freeplane. I think you might know the better places or directions for this kind of features.
> > As for me, I think #202 is too much hard-coded so I hesitated and closed #202 .
> 
> Well, actually it was hard-coded because the (pretty old) original code is also hard-coding this logic.
> And I do not apply the same high code quality criteria to the old code as to my newest and latest projects. Because life is too short to refactor everything.
> 
> So I would actually suggest that we have a conversation about it. If you prefer we can continue to discuss everything here, but I would also be interested to have video chat with screen sharing. I am generally available on Fridays, Saturdays and Sundays, if it sounds good for you.

I created a sourceforge account so we can discuss there!

dpolivaev commented:
> I created a sourceforge account so we can discuss there!

Whatever you prefer, I am available when you need me in the way you prefer.

I would like to tell you one analogy. Developing software is like playing golf. You almost always need multiple strokes to reach the hole. And here the position of the hole is not as well defined and should be discovered as we go.  And personally, I enjoy playing it together doing the strokes one after very much.

danix800 commented:
> > I created a sourceforge account so we can discuss there!
> 
> Whatever you prefer, I am available when you need me in the way you prefer.
> 
> I would like to tell you one analogy. Developing software is like playing golf. You almost always need multiple strokes to reach the hole. And here the position of the hole is not as well defined and should be discovered as we go. And personally, I enjoy playing it together doing the strokes one after very much.

I agree! Rome is not built in one day. I really appreciate your nice support and understanding and I'll keep on contributing as much as I can.
